src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -813,4 +823,25 @@ void BaseCouplingScheme::receiveConvergence(const m2n::PtrM2N &m2n)
   m2n->receive(_hasConverged);
 }
 
+void BaseCouplingScheme::addTimeStepAtWindowEnd()
+{
+
+  for (auto &data : _allData | boost::adaptors::map_values) {
+    auto   timeStepStorage = data->timeStepsStorage();
+    double lastSampleT     = timeStepStorage.maxStoredTime();
+    double endOfTimeWindow = getWindowEndTime();
+    //If there does not exist a sample at the window end duplicate the last available sample and put it at the end of the time window.
+    if (lastSampleT < endOfTimeWindow - math::NUMERICAL_ZERO_DIFFERENCE) {
+      data->setSampleAtTime(endOfTimeWindow, timeStepStorage.getSampleAtOrAfter(lastSampleT));
+    }
+    //This keeps the time within preCICE synchronized
+    _computedTimeWindowPart = _timeWindowSize;
```

Comment:
This line should be outside of the for-loop (does not make sense to call it several times). Why not even put it outside of `BaseCouplingScheme::addTimeStepAtWindowEnd`? Does not really belong here in my opinion.